react-custom-hooks/client-ts/src/components/useTodos.ts

```diff
@@ -0,0 +1,135 @@
+/* eslint-disable @typescript-eslint/no-unused-vars -- Remove me */
+import { useEffect, useState } from 'react';
+
+export type UnsavedTodo = {
+  task: string;
+  isCompleted: boolean;
+};
+export type Todo = UnsavedTodo & {
+  todoId: number;
+};
+
+export type UseTodosValues = {
+  isLoading: boolean;
+  error?: unknown;
+  todos?: Todo[];
+  addTodo: (todo: UnsavedTodo) => Promise<void>;
+  toggleCompleted: (todoId: number) => Promise<void>;
+};
+/**
+ * Manages the Todos by reading from and writing to the backend API,
+ * using the API functions below.
+ */
+export function useTodos(): UseTodosValues {
+  const [todos, setTodos] = useState<Todo[]>();
+  const [error, setError] = useState<unknown>();
+  useEffect(() => {
+    /* TODO: If `todos` hasn't been defined yet, read the items from the API
+     * and set them into the `todos` state.
+     */
+    async function fetchTodos() {
+      try {
+        const todosData = await readTodos();
+        setTodos(todosData);
+      } catch (error) {
+        setError(error);
+      }
+    }
+    if (!todos) {
+      fetchTodos();
+    }
+  }, [todos]);
+
+  async function addTodo(newTodo: UnsavedTodo): Promise<void> {
+    /* TODO: Call the API function that creates a Todo item.
+     * When the promise returned by that function resolves, update the `todos` state.
+     * Note that it is critical that you pass a _new_ array. Do not modify the `todos` array.
+     */
+    try {
+      const createdTodo = await createTodo(newTodo);
+      const updatedTodos = [...(todos ?? []), createdTodo];
+      setTodos(updatedTodos);
+    } catch (error) {
+      setError(error);
+    }
+  }
+
+  async function toggleCompleted(todoId: number): Promise<void> {
+    /* TODO: Find the Todo item being updated, toggle its completed prop, and call
+     * the API function that updates a Todo item.
+     * When the promise returned by that function resolves, update the `todos` state.
+     * When updating this state, use the updated `todo` returned from the API.
+     * Note that it is critical that you pass a _new_ array. Do not modify the `todos` array.
+     */
+    try {
+      if (!todos) {
```

Comment:
Oh, you did that on this function! Up above I should have just said, "Do what you did down below." :-)